File: simple_factory.cpp
#include <iostream>
using namespace std;



// Shape interface
class Shape {
public:
    virtual void draw() = 0; // Pure virtual function
    virtual ~Shape() {} // Virtual destructor
};


// Circle class implementing Shape interface
class Circle : public Shape {
public:
    void draw() override {
        cout << "Drawing a Circle" << endl;
    }
};

// Square class implementing Shape interface
class Square : public Shape {
public:
    void draw() override {
        cout << "Drawing a Square" << endl;
    }
};

// ShapeFactory class to create shapes
class ShapeFactory {
public:
    // Factory method to create shapes
    static Shape* createShape(const string& shapeType) {
        if (shapeType == "Circle") {
            return new Circle();
        } else if (shapeType == "Square") {
            return new Square();
        }
        return nullptr; // Return null if shape type is unknown
    }
};

int main() {
    // Create a Circle using the factory
    Shape* shape1 = ShapeFactory::createShape("Circle");
    shape1->draw(); // Output: Drawing a Circle

    // Create a Square using the factory
    Shape* shape2 = ShapeFactory::createShape("Square");
    shape2->draw(); // Output: Drawing a Square

    // Clean up memory
    delete shape1;
    delete shape2;

    return 0;
}

File: design_patterns/typescript/simple_factory.ts
// Shape interface
interface Shape {
    draw(): void;
}

// Circle class implementing Shape interface
class Circle implements Shape {
    draw(): void {
        console.log("Drawing a Circle");
    }
}

// Square class implementing Shape interface
class Square implements Shape {
    draw(): void {
        console.log("Drawing a Square");
    }
}

// ShapeFactory class to create shapes
class ShapeFactory {
    // Registry of shape types and their constructors
    private static readonly shapeMap = new Map<string, new () => Shape>();

    /**
     * Register a new shape type (allows extension without modification - OCP compliant)
     */
    static register(shapeType: string, shapeClass: new () => Shape): void {
        this.shapeMap.set(shapeType, shapeClass);
    }

    /**
     * Factory method to create shapes using a Map registry.
     */
    static createShape(shapeType: string): Shape | null {
        const ShapeClass = this.shapeMap.get(shapeType);
        return ShapeClass ? new ShapeClass() : null;
    }
}

// Register shapes (can be done from anywhere without modifying ShapeFactory)
ShapeFactory.register("Circle", Circle);
ShapeFactory.register("Square", Square);

// Main function
function main(): void {
    // Create a Circle using the factory
    const shape1 = ShapeFactory.createShape("Circle");
    if (shape1) shape1.draw(); // Output: Drawing a Circle

    // Create a Square using the factory
    const shape2 = ShapeFactory.createShape("Square");
    if (shape2) shape2.draw(); // Output: Drawing a Square
}

// Execute main
main();